File: desktop/src-tauri/src/env.rs
// Environment overrides for the agent + auth sidecars.
//
// - ACCOUNTANT24_HOME / PI_CODING_AGENT_DIR point pi at the user's workspace
//   (~/Accountant24) so the ledger, auth.json, and models.json live together — what
//   the auth helper writes is what the agent reads.
// - PI_PACKAGE_DIR points stock pi at its staged sibling assets (package.json,
//   theme, export-html) since the sidecar's execPath isn't next to them.
// - PATH / TESSDATA_PREFIX expose the bundled native tools (hledger, pdftotext,
//   tesseract). Launched from Finder, PATH is minimal and excludes Homebrew —
//   bundling + this injection is what makes it zero-setup.
//
// In `tauri dev` the resources aren't bundled into the app, so we resolve them
// from the source tree (populated by `bun run build`); see agent_resource_dir.

use std::collections::HashMap;
use std::path::PathBuf;
use tauri::{AppHandle, Manager};

/// Directory holding the staged agent resources (the extension JS, the `pi/` asset
/// dir, and bundled `bin/`/`tessdata/`). In a release bundle this is the app's
/// resource dir; in `tauri dev` resources aren't bundled, so we fall back to the
/// source-tree staging location (`desktop/src-tauri/`) populated by `bun run build`.
pub fn agent_resource_dir(app: &AppHandle) -> Option<PathBuf> {
    if cfg!(debug_assertions) {
        Some(PathBuf::from(env!("CARGO_MANIFEST_DIR")))
    } else {
        app.path().resource_dir().ok()
    }
}

/// Absolute path to the bundled extension JS passed to `pi -e`.
pub fn extension_path(app: &AppHandle) -> Option<PathBuf> {
    agent_resource_dir(app).map(|d| d.join("accountant24-extension.js"))
}

/// Env overrides shared by the agent + auth sidecars.
pub fn sidecar_env(app: &AppHandle) -> HashMap<String, String> {
    let mut env = HashMap::new();

    // Workspace dir: ledger + auth.json + models.json all under ~/Accountant24.
    if let Ok(home) = app.path().home_dir() {
        let workspace = home.join("Accountant24").to_string_lossy().to_string();
        env.insert("ACCOUNTANT24_HOME".to_string(), workspace.clone());
        env.insert("PI_CODING_AGENT_DIR".to_string(), workspace);
    }

    let Some(resource_dir) = agent_resource_dir(app) else {
        return env;
    };

    // pi reads its package.json / theme / export-html from PI_PACKAGE_DIR.
    let pi_dir = resource_dir.join("pi-assets");
    if pi_dir.is_dir() {
        env.insert(
            "PI_PACKAGE_DIR".to_string(),
            pi_dir.to_string_lossy().to_string(),
        );
    }

    let bin = resource_dir.join("bin");
    if bin.is_dir() {
        let bin_str = bin.to_string_lossy().to_string();
        let current = std::env::var("PATH").unwrap_or_default();
        let path = if current.is_empty() {
            bin_str
        } else {
            format!("{bin_str}:{current}")
        };
        env.insert("PATH".to_string(), path);
    }

    let tessdata = resource_dir.join("tessdata");
    if tessdata.is_dir() {
        // tesseract 4/5 resolves `<lang>.traineddata` directly under TESSDATA_PREFIX.
        env.insert(
            "TESSDATA_PREFIX".to_string(),
            tessdata.to_string_lossy().to_string(),
        );
    }

    env
}

File: packages/desktop/electron/main/env.ts
// Resource paths + the environment passed to the pi agent child.
//
// Mirrors the old src-tauri/src/env.rs contract: the workspace (~/Accountant24)
// holds the ledger + auth.json + models.json; PATH exposes the vendored native
// tools (hledger/pdftotext/tesseract); TESSDATA_PREFIX points at the OCR data.
// pi runs from node_modules, so PI_PACKAGE_DIR is no longer needed (its assets
// sit beside it) — add it back here only if a future pi version requires it.

import { existsSync } from "node:fs";
import { homedir } from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { app } from "electron";

/** ~/Accountant24 — the agent's cwd and the home of the ledger/auth/models. */
export function workspaceDir(): string {
  const env = process.env.ACCOUNTANT24_HOME;
  return env && env.length > 0 ? env : path.join(homedir(), "Accountant24");
}

/** ~/Accountant24/skills — one self-contained folder per installed skill
 *  (Agent Skills standard: a dir with SKILL.md). Each enabled skill is passed
 *  to the agent child via a `--skill` flag. */
export function skillsDir(): string {
  return path.join(workspaceDir(), "skills");
}

/** ~/Accountant24/sessions — pi's session files, one JSONL per chat thread.
 *  Passed to the agent child via `--session-dir`. */
export function sessionsDir(): string {
  return path.join(workspaceDir(), "sessions");
}

/** ~/Accountant24/ledger/main.journal — the ledger's entry point (includes the
 *  other journal files). */
export function mainJournalPath(): string {
  return path.join(workspaceDir(), "ledger", "main.journal");
}

/** ~/Accountant24/app-settings.json — app-owned settings (distinct from pi's). */
export function appSettingsPath(): string {
  return path.join(workspaceDir(), "app-settings.json");
}

/** ~/Accountant24/settings.json — pi's settings file, which earlier app
 *  versions shared; read once as a migration source. */
export function legacySettingsPath(): string {
  return path.join(workspaceDir(), "settings.json");
}

/** Dir holding vendored bin/ + tessdata/ + the extension bundle.
 *  Dev: packages/desktop/resources. Packaged: the app's resources dir
 *  (electron-builder extraResources land directly under it). */
function resourceDir(): string {
  return app.isPackaged ? process.resourcesPath : path.join(app.getAppPath(), "resources");
}

/** Absolute path to pi's Node CLI entry (run with Electron-as-Node). pi is
 *  ESM-only (its package "exports" has no `require` condition), so resolve via
 *  import.meta.resolve (the "." entry → dist/index.js) and take sibling cli.js. */
export function piCliPath(): string {
  const main = fileURLToPath(import.meta.resolve("@earendil-works/pi-coding-agent"));
  return path.join(path.dirname(main), "cli.js");
}

/** Dir holding the vendored native tools (hledger/pdftotext/tesseract). Prepended
 *  to the agent child's PATH; also used to resolve a tool's absolute path when we
 *  run one directly from the main process (which does NOT inherit that PATH). */
export function binDir(): string {
  return path.join(resourceDir(), "bin");
}

/** The bundled extension passed to `pi -e`. Loaded as JS in both dev and
 *  packaged (Electron-as-Node can't parse the TS source); produced by
 *  scripts/bundle-extension.ts. */
export function extensionPath(): string {
  return path.join(resourceDir(), "accountant24-extension.js");
}

/** The static system prompt passed to `pi --system-prompt`. pi replaces its
 *  coding-agent preamble with this file's contents but still assembles its
 *  native sections (the <available_skills> block, date/cwd) around it; the
 *  extension then appends the dynamic tools/context sections per turn. Copied
 *  next to the extension bundle by scripts/bundle-extension.ts. */
export function systemPromptPath(): string {
  return path.join(resourceDir(), "system.md");
}

/** Native (built-in) skills embedded in the app bundle — one folder per skill,
 *  committed under packages/desktop/resources/skills. Always loaded (a single
 *  `--skill` flag; pi recurses the directory), never present in the workspace
 *  skills folder, so users can't remove or disable them. */
export function nativeSkillsDir(): string {
  return path.join(resourceDir(), "skills");
}

/** Binary for running pi as Node (ELECTRON_RUN_AS_NODE). On macOS, use the
 *  bundled Helper app's binary instead of the main app binary: pi touches
 *  AppKit at startup, which makes LaunchServices register the child under its
 *  bundle's Info.plist — for the main binary that's a regular app, i.e. a
 *  second (generic "exec") Dock icon. The helper bundles are LSUIElement, so
 *  the same process stays invisible. Falls back to the main binary if the
 *  helper isn't at the expected path. */
export function nodeRuntimePath(): string {
  if (process.platform !== "darwin") return process.execPath;
  // <App>.app/Contents/MacOS/<App> -> <App>.app (also matches dev's Electron.app)
  const bundle = path.resolve(process.execPath, "..", "..", "..");
  const name = path.basename(bundle, ".app");
  const helper = path.join(
    bundle,
    "Contents",
    "Frameworks",
    `${name} Helper (Plugin).app`,
    "Contents",
    "MacOS",
    `${name} Helper (Plugin)`,
  );
  return existsSync(helper) ? helper : process.execPath;
}

/** Env overrides for the pi child + in-process SDK: workspace + vendored tools. */
export function agentEnv(): NodeJS.ProcessEnv {
  const workspace = workspaceDir();
  const env: NodeJS.ProcessEnv = {
    ...process.env,
    ACCOUNTANT24_HOME: workspace,
    PI_CODING_AGENT_DIR: workspace,
  };
  const res = resourceDir();
  const bin = binDir();
  if (existsSync(bin)) env.PATH = `${bin}${path.delimiter}${env.PATH ?? ""}`;
  const tessdata = path.join(res, "tessdata");
  if (existsSync(tessdata)) env.TESSDATA_PREFIX = tessdata;
  return env;
}
